File: src/main/java/com/pauljoda/modularsystems/storage/tiles/TileStorageSmashing.java
package com.pauljoda.modularsystems.storage.tiles;

import com.pauljoda.modularsystems.storage.container.ContainerStorageSmashing;
import com.pauljoda.modularsystems.storage.gui.GuiStorageSmashing;
import com.teambr.bookshelf.collections.InventoryTile;
import com.teambr.bookshelf.collections.Location;
import com.teambr.bookshelf.common.blocks.rotation.SixWayRotation;
import net.minecraft.block.Block;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.inventory.IInventory;
import net.minecraft.item.ItemPickaxe;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import net.minecraft.util.StatCollector;
import net.minecraft.world.World;
import net.minecraftforge.common.util.ForgeDirection;

import java.util.ArrayList;

/**
 * Modular-Systems
 * Created by Paul Davis on 7/27/2015
 */
public class TileStorageSmashing extends TileEntityStorageExpansion implements IInventory {

    protected InventoryTile pickSlot;

    public TileStorageSmashing() {
        pickSlot = new InventoryTile(1);
    }

    @Override
    public void addedToNetwork() {}

    @Override
    public void removedFromNetwork() {
        ItemStack itemStack = pickSlot.getStackInSlot(0);

        if (itemStack != null && itemStack.stackSize > 0) {
            float rx = worldObj.rand.nextFloat() * 0.8F + 0.1F;
            float ry = worldObj.rand.nextFloat() * 0.8F + 0.1F;
            float rz = worldObj.rand.nextFloat() * 0.8F + 0.1F;

            EntityItem entityItem = new EntityItem(worldObj,
                    xCoord + rx, yCoord + ry, zCoord + rz,
                    new ItemStack(itemStack.getItem(), itemStack.stackSize, itemStack.getItemDamage()));

            if (itemStack.hasTagCompound())
                entityItem.getEntityItem().setTagCompound((NBTTagCompound) itemStack.getTagCompound().copy());

            float factor = 0.05F;
            entityItem.motionX = worldObj.rand.nextGaussian() * factor;
            entityItem.motionY = worldObj.rand.nextGaussian() * factor + 0.2F;
            entityItem.motionZ = worldObj.rand.nextGaussian() * factor;
            worldObj.spawnEntityInWorld(entityItem);

            itemStack.stackSize = 0;
        }
    }

    /*******************************************************************************************************************
     *********************************************** Tile Methods ******************************************************
     *******************************************************************************************************************/

    @Override
    @SuppressWarnings("unchecked")
    public void updateEntity() {
        super.updateEntity();
        if(!worldObj.isRemote && !isPowered() && getCore() != null && pickSlot.getStackInSlot(0) != null && worldObj.rand.nextInt(20) == 0) {
            SixWayRotation rotation = new SixWayRotation();
            ForgeDirection dir = rotation.convertMetaToDirection(worldObj.getBlockMetadata(xCoord, yCoord, zCoord));
            Location blockBreakLocation = new Location(this);
            blockBreakLocation.travel(dir);
            if(!worldObj.isAirBlock(blockBreakLocation.x, blockBreakLocation.y, blockBreakLocation.z)) {
                Block toBreak = worldObj.getBlock(blockBreakLocation.x, blockBreakLocation.y, blockBreakLocation.z);
                int harvestLevel = pickSlot.getStackInSlot(0).getItem().getHarvestLevel(pickSlot.getStackInSlot(0), "pickaxe");
                if(toBreak.getHarvestLevel(worldObj.getBlockMetadata(blockBreakLocation.x, blockBreakLocation.y, blockBreakLocation.z)) <= harvestLevel) {

                    int fortune = 0;

                    NBTTagList enchantList = pickSlot.getStackInSlot(0).getEnchantmentTagList();
                    if(enchantList != null) {
                        for (int i = 0; i < enchantList.tagCount(); i++) {
                            NBTTagCompound tag = enchantList.getCompoundTagAt(0);
                            if (tag.hasKey("id") && tag.getInteger("id") == 35)
                                fortune = tag.getInteger("lvl");
                        }
                    }

                    ArrayList<ItemStack> itemStacks = toBreak.getDrops(worldObj, blockBreakLocation.x, blockBreakLocation.y, blockBreakLocation.z,
                            worldObj.getBlockMetadata(blockBreakLocation.x, blockBreakLocation.y, blockBreakLocation.z), fortune);

                    for(ItemStack itemStack : itemStacks) {
                        if (itemStack != null && itemStack.stackSize > 0) {
                            float rx = worldObj.rand.nextFloat() * 0.8F + 0.1F;
                            float ry = worldObj.rand.nextFloat() * 0.8F + 0.1F;
                            float rz = worldObj.rand.nextFloat() * 0.8F + 0.1F;

                            EntityItem entityItem = new EntityItem(worldObj,
                                    blockBreakLocation.x + rx, blockBreakLocation.y + ry, blockBreakLocation.z + rz,
                                    new ItemStack(itemStack.getItem(), itemStack.stackSize, itemStack.getItemDamage()));

                            if (itemStack.hasTagCompound())
                                entityItem.getEntityItem().setTagCompound((NBTTagCompound) itemStack.getTagCompound().copy());

                            float factor = 0.05F;
                            entityItem.motionX = worldObj.rand.nextGaussian() * factor;
                            entityItem.motionY = worldObj.rand.nextGaussian() * factor + 0.2F;
                            entityItem.motionZ = worldObj.rand.nextGaussian() * factor;
                            worldObj.spawnEntityInWorld(entityItem);

                            itemStack.stackSize = 0;
                        }
                    }
                    worldObj.playAuxSFX(2001, blockBreakLocation.x, blockBreakLocation.y, blockBreakLocation.z, Block.getIdFromBlock(toBreak) + ( worldObj.getBlockMetadata(blockBreakLocation.x, blockBreakLocation.y, blockBreakLocation.z) << 12));
                    worldObj.setBlockToAir(blockBreakLocation.x, blockBreakLocation.y, blockBreakLocation.z);
                    pickSlot.getStackInSlot(0).setItemDamage(pickSlot.getStackInSlot(0).getItemDamage() + 1);
                    if(pickSlot.getStackInSlot(0).getItemDamage() > pickSlot.getStackInSlot(0).getMaxDamage())
                        pickSlot.setStackInSlot(null, 0);
                }
            }
        }
    }

    @Override
    public void readFromNBT (NBTTagCompound tag) {
        super.readFromNBT(tag);
        pickSlot.readFromNBT(tag);
    }

    @Override
    public void writeToNBT (NBTTagCompound tag) {
        super.writeToNBT(tag);
        pickSlot.writeToNBT(tag);
    }

    /*******************************************************************************************************************
     ***************************************** IInventory Methods ******************************************************
     *******************************************************************************************************************/

    @Override
    public int getSizeInventory() {
        return 1;
    }

    @Override
    public ItemStack getStackInSlot(int slot) {
        return pickSlot.getStackInSlot(slot);
    }

    @Override
    public ItemStack decrStackSize(int slot, int amount) {
        if(pickSlot.getStackInSlot(slot) != null) {
            ItemStack returnStack;
            if(pickSlot.getStackInSlot(slot).stackSize <= amount) {
                returnStack = pickSlot.getStackInSlot(slot);
                pickSlot.setStackInSlot(null, slot);
                this.markDirty();
                return returnStack;
            } else {
                returnStack = pickSlot.getStackInSlot(slot).splitStack(amount);
                if(pickSlot.getStackInSlot(slot).stackSize <= 0)
                    pickSlot.setStackInSlot(null, slot);
                this.markDirty();
                return returnStack;
            }
        }
        return null;
    }

    @Override
    public ItemStack getStackInSlotOnClosing(int slot) {
        return pickSlot.getStackInSlot(slot);
    }

    @Override
    public void setInventorySlotContents(int slot, ItemStack stack) {
        pickSlot.setStackInSlot(stack, slot);
    }

    /**
     * Returns the name of the inventory
     */
    public String getInventoryName() {
        return StatCollector.translateToLocal("inventory.storageSmashing.title");
    }

    /**
     * Returns if the inventory is named
     */
    public boolean hasCustomInventoryName() {
        return true;
    }

    @Override
    public int getInventoryStackLimit() {
        return 1;
    }

    /**
     * Do not give this method the name canInteractWith because it clashes with Container
     */
    public boolean isUseableByPlayer(EntityPlayer player) {
        return getCore() != null;
    }

    @Override
    public void openInventory() {}

    @Override
    public void closeInventory() {}

    @Override
    public boolean isItemValidForSlot(int slot, ItemStack stack) {
        return stack.getItem() instanceof ItemPickaxe;
    }

    /*******************************************************************************************************************
     ****************************************** IOpensGui Methods ******************************************************
     *******************************************************************************************************************/

    @Override
    public Object getServerGuiElement(int ID, EntityPlayer player, World world, int x, int y, int z) {
        return getCore() != null ? new ContainerStorageSmashing(player.inventory, this) : null;
    }

    @Override
    public Object getClientGuiElement(int ID, EntityPlayer player, World world, int x, int y, int z) {
        return getCore() != null ? new GuiStorageSmashing(new ContainerStorageSmashing(player.inventory, this)) : null;
    }
}

File: src/main/scala/com/teambr/modularsystems/storage/tiles/TileStorageSmashing.scala
package com.teambr.modularsystems.storage.tiles

import java.util

import com.teambr.bookshelf.collections.Location
import com.teambr.bookshelf.common.blocks.properties.PropertyRotation
import com.teambr.bookshelf.common.tiles.traits.{ RedstoneAware, Inventory }
import net.minecraft.block.Block
import net.minecraft.block.material.Material
import net.minecraft.entity.item.EntityItem
import net.minecraft.item.ItemStack
import net.minecraft.nbt.{ NBTTagCompound, NBTTagList }
import net.minecraft.util.EnumFacing

/**
 * This file was created for Modular-Systems
 *
 * Modular-Systems is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Paul Davis pauljoda
 * @since August 14, 2015
 */
class TileStorageSmashing extends TileEntityStorageExpansion with Inventory with RedstoneAware {
    /**
     * Called after this has been added to a network
     */
    override def addedToNetwork() : Unit = {}

    /**
     * Called right before this is removed from a network
     */
    override def removedFromNetwork() : Unit = {
        val itemStack : ItemStack = getStackInSlot(0)

        if (itemStack != null && itemStack.stackSize > 0) {
            val rx : Float = worldObj.rand.nextFloat * 0.8F + 0.1F
            val ry : Float = worldObj.rand.nextFloat * 0.8F + 0.1F
            val rz : Float = worldObj.rand.nextFloat * 0.8F + 0.1F
            val entityItem : EntityItem = new EntityItem(worldObj, pos.getX + rx, pos.getY + ry, pos.getZ + rz, new ItemStack(itemStack.getItem, itemStack.stackSize, itemStack.getItemDamage))
            if (itemStack.hasTagCompound) entityItem.getEntityItem.setTagCompound(itemStack.getTagCompound.copy.asInstanceOf[NBTTagCompound])
            val factor : Float = 0.05F
            entityItem.motionX = worldObj.rand.nextGaussian * factor
            entityItem.motionY = worldObj.rand.nextGaussian * factor + 0.2F
            entityItem.motionZ = worldObj.rand.nextGaussian * factor
            worldObj.spawnEntityInWorld(entityItem)
            itemStack.stackSize = 0
        }
    }

    override def update() : Unit = {
        super.update()
        if (!worldObj.isRemote && !isPowered && getCore != null && getStackInSlot(0) != null && worldObj.rand.nextInt(20) == 0) {

            val dir : EnumFacing = worldObj.getBlockState(pos).getValue(PropertyRotation.SIX_WAY.getProperty).asInstanceOf[EnumFacing]
            val blockBreakLocation = new Location(pos)
            blockBreakLocation.travel(dir)

            if (!worldObj.isAirBlock(blockBreakLocation.asBlockPos)) {

                val toBreak : Block = worldObj.getBlockState(blockBreakLocation.asBlockPos).getBlock
                if (toBreak.getMaterial == Material.water || toBreak.getMaterial == Material.lava) return
                val harvestLevel : Int = getStackInSlot(0).getItem.getHarvestLevel(getStackInSlot(0), "pickaxe")

                //Can we harvest this?
                if (toBreak.getHarvestLevel(worldObj.getBlockState(blockBreakLocation.asBlockPos)) <= harvestLevel) {
                    var fortune : Int = 0
                    val silkTouch : Boolean = toBreak.canSilkHarvest(worldObj, blockBreakLocation.asBlockPos, worldObj.getBlockState(blockBreakLocation.asBlockPos), null)
                    var hasSilkTouch : Boolean = false
                    val enchantList : NBTTagList = getStackInSlot(0).getEnchantmentTagList

                    //Check for enchants
                    if (enchantList != null) {
                        for(i <- 0 until enchantList.tagCount()) {
                            val tag = enchantList.getCompoundTagAt(i)
                            if (tag.hasKey("id") && tag.getInteger("id") == 35) fortune = tag.getInteger("lvl")
                            else if (tag.hasKey("id") && tag.getInteger("id") == 33) hasSilkTouch = true
                        }
                    }

                    //Build Drop List
                    val itemStacks : util.ArrayList[ItemStack] = new util.ArrayList[ItemStack]
                    if (silkTouch && hasSilkTouch) {
                        itemStacks.add(new ItemStack(toBreak, 1, worldObj.getBlockState(blockBreakLocation.asBlockPos).getBlock.getMetaFromState(worldObj.getBlockState(blockBreakLocation.asBlockPos))))
                    }
                    else {
                        itemStacks.addAll(toBreak.getDrops(worldObj, blockBreakLocation.asBlockPos, worldObj.getBlockState(blockBreakLocation.asBlockPos), fortune))
                    }

                    //Spawn Drop List
                    for (i <- 0 until itemStacks.size()) {
                        val itemStack = itemStacks.get(i)
                        if (itemStack != null && itemStack.stackSize > 0) {
                            val rx : Float = worldObj.rand.nextFloat * 0.8F + 0.1F
                            val ry : Float = worldObj.rand.nextFloat * 0.8F + 0.1F
                            val rz : Float = worldObj.rand.nextFloat * 0.8F + 0.1F
                            val entityItem : EntityItem = new EntityItem(worldObj, blockBreakLocation.x + rx, blockBreakLocation.y + ry, blockBreakLocation.z + rz, new ItemStack(itemStack.getItem, itemStack.stackSize, itemStack.getItemDamage))
                            if (itemStack.hasTagCompound) entityItem.getEntityItem.setTagCompound(itemStack.getTagCompound.copy.asInstanceOf[NBTTagCompound])
                            val factor : Float = 0.05F
                            entityItem.motionX = worldObj.rand.nextGaussian * factor
                            entityItem.motionY = worldObj.rand.nextGaussian * factor + 0.2F
                            entityItem.motionZ = worldObj.rand.nextGaussian * factor
                            worldObj.spawnEntityInWorld(entityItem)
                            itemStack.stackSize = 0
                        }
                    }

                    //Destroy Block
                    worldObj.destroyBlock(blockBreakLocation.asBlockPos, false)
                    getStackInSlot(0).setItemDamage(getStackInSlot(0).getItemDamage + 1)
                    if (getStackInSlot(0).getItemDamage > getStackInSlot(0).getMaxDamage) setInventorySlotContents(0, null)
                }
            }
        }
    }

    override def markDirty() : Unit = {
        super[TileEntityStorageExpansion].markDirty()
        super[Inventory].markDirty()
    }

    override def readFromNBT(tag : NBTTagCompound) : Unit = {
        super[TileEntityStorageExpansion].readFromNBT(tag)
        super[Inventory].readFromNBT(tag)
    }

    override def writeToNBT(tag : NBTTagCompound) : Unit = {
        super[TileEntityStorageExpansion].writeToNBT(tag)
        super[Inventory].writeToNBT(tag)
    }

    /*******************************************************************************************************************
      ***************************************** Inventory Methods ******************************************************
      *******************************************************************************************************************/

    override var inventoryName : String = "inventory.storageSmashing.title"

    override def hasCustomName() : Boolean = true

    override def initialSize : Int = 1
}
